builders_guide.md

```diff
@@ -0,0 +1,547 @@
+        
+        ```jsx
+        const UPGRADEABLE_PLANETS = uiManager.getGameManager().getContractConstants().UPGRADEABLE_PLANETS;
+        ```
+        
+    - SPOILER
+        
+        ```typescript
+        // ~ Line 88
+        const UPGRADEABLE_PLANETS = uiManager.getGameManager().getContractConstants().UPGRADEABLE_PLANETS;
+        
+          if (planet && account) {
+            if (planet.owner !== account) {
+            } else if (!UPGRADEABLE_PLANETS[planet.planetType] || planet.silverCap === 0) {
+              content = (
+                <CenterBackgroundSubtext width={'100%'} height='100px'>
+                  This Planet <br /> is not Upgradeable
+                </CenterBackgroundSubtext>
+              );
+            }
+        ```
+        
+
+## 3.8 Run a local game.
+
+- `yarn` to compile the client side changes
+- `yarn start` to launch the game. (Same as step 1.1)
+    - *Note: For test purposes, you can set `DISABLE_ZK_CHECKS` to true for the local game.*
+- See if you can upgrade an asteroid field!
+    - [https://twitter.com/cha0sg0d_/status/1484620387866447873](https://twitter.com/cha0sg0d_/status/1484620387866447873)
+
+A complete solution to Quest 3 can be found on the tutorial branch of [cha0sg0d/eth](https://github.com/cha0sg0d/eth) and [cha0sg0d/client.](https://github.com/cha0sg0d/client)
+
+# Quest 4: Deploy Contracts and Client to Production.
+
+Follow the steps in the [README](README.md#static-deployment-of-dark-forest-no-webserver) to deploy the contracts and client.
+
+If your round includes custom planets, run the following command after deploying DarkForestCore:
+
+### 4.1 Add custom planet (optional).
+
+- `yarn workspace eth hardhat:prod game:createPlanets`
+    - This functionality is built-in to the local deploy process, but needs to happen manually for deploying to production.
+
+
+## Side Quest 1: Deploying the subgraph
+
+- What is [The Graph](https://thegraph.com/) and why is it useful?
+    - [The Graph](https://thegraph.com/) is an indexing protocol for querying networks like Ethereum and IPFS. Anyone can build and publish open APIs, called subgraphs, making data easily accessible.
+    - The graph enables more complex queries like who owned this planet at time *t* and has a much faster performance than a vanilla JSON-RPC request.
+- The amazing Jacob Rosenthal has created a DF subgraph that can be deployed via just a few commands. Follow his instructions [here](https://github.com/darkforest-eth/eth/tree/master/subgraph) to deploy the subgraph.
+    - *Note: If you’re on a Mac M1, Docker might not work, and you won’t be able to test the subgraph locally.*
+- If you are changing a fundamental data structure of the subgraph, you will need to update the schema and mappings. See [here](https://github.com/cha0sg0d/eth/commit/1b382624c986ed2b045b8465bde2df543dc5d3f0) for an example edit.
+
+## Side Quest 2: Auto Whitelisting with no server.
+
+*Warning: This information is potentially out of date*
```

Comment:
```suggestion
```

It was, but let's fix it.

Reply:
I will link to my Whitelist.sol and hardhat test if people want to set up no server whitelisting.